lib/ex_webrtc/dtls_transport.ex

```diff
@@ -1,148 +1,239 @@
+    GenServer.call(dtls_transport, :get_fingerprint)
+  end
+
+  @doc false
+  @spec start_dtls(dtls_transport(), :active | :passive) :: :ok | {:error, :already_started}
+  def start_dtls(dtls_transport, mode) do
+    GenServer.call(dtls_transport, {:start_dtls, mode})
+  end
+
+  @doc false
+  @spec send_data(dtls_transport(), binary()) :: :ok
+  def send_data(dtls_transport, data) do
+    GenServer.cast(dtls_transport, {:send_data, data})
+  end
+
+  @impl true
+  def init([ice_config, peer_connection]) do
     # temporary hack to generate certs
     {:ok, cert_client} = ExDTLS.start_link(client_mode: true, dtls_srtp: true)
     {:ok, cert} = ExDTLS.get_cert(cert_client)
     {:ok, pkey} = ExDTLS.get_pkey(cert_client)
     {:ok, fingerprint} = ExDTLS.get_cert_fingerprint(cert_client)
     :ok = ExDTLS.stop(cert_client)
 
+    {:ok, ice_agent} = ICEAgent.start_link(:controlled, ice_config)
     srtp = ExLibSRTP.new()
 
-    %__MODULE__{
+    state = %{
+      peer_connection: peer_connection,
       ice_agent: ice_agent,
+      ice_state: nil,
+      buffered_packets: nil,
       cert: cert,
       pkey: pkey,
       fingerprint: fingerprint,
-      srtp: srtp
+      srtp: srtp,
+      dtls_state: :new
     }
+
+    {:ok, state}
   end
 
-  def start(dtls, :passive) do
-    {:ok, client} =
-      ExDTLS.start_link(
-        client_mode: false,
-        dtls_srtp: true,
-        pkey: dtls.pkey,
-        cert: dtls.cert
-      )
+  @impl true
+  def handle_call(:get_ice_agent, _from, state) do
+    {:reply, state.ice_agent, state}
+  end
+
+  @impl true
+  def handle_call(:get_fingerprint, _from, state) do
+    {:reply, state.fingerprint, state}
+  end
 
-    %__MODULE__{dtls | client: client, mode: :passive}
+  @impl true
+  def handle_call({:start_dtls, _mode}, _from, %{client: _client} = state) do
+    # is there a case when mode will change and new handshake will be needed?
+    {:reply, {:error, :already_started}, state}
   end
 
-  def start(dtls, :active) do
+  @impl true
+  def handle_call({:start_dtls, mode}, _from, state) when mode in [:active, :passive] do
     {:ok, client} =
       ExDTLS.start_link(
-        client_mode: true,
+        client_mode: mode == :active,
         dtls_srtp: true,
-        pkey: dtls.pkey,
-        cert: dtls.cert
+        pkey: state.pkey,
+        cert: state.cert
       )
 
-    # we assume that ICE in not in connected state yet
-    %__MODULE__{dtls | client: client, mode: :active}
+    state =
+      state
+      |> Map.put(:client, client)
+      |> Map.put(:mode, mode)
+
+    {:reply, :ok, state}
   end
 
-  def update_ice_state(dtls, :connected) do
-    if dtls.mode == :active do
-      {:ok, packets} = ExDTLS.do_handshake(dtls.client)
-      :ok = ICEAgent.send_data(dtls.ice_agent, packets)
-    end
+  @impl true
+  def handle_cast({:send_data, _data}, %{dtls_state: :connected, ice_state: ice_state} = state)
+      when ice_state in [:connected, :completed] do
+    # TODO
```

Comment:
It would be nice to encrypt the data firstly :) Also, I imagine that the function will receive a binary, so RTP/RTCP after encoding instead of the RTP/RTCP struct, but that might be a little bit inconsistent if we want to do the demuxing in `DTLSTransport`, hard to tell if that's an issue.